### ISSUE TITLE
fetech_pandas_all for large timestamp get out of bound timestmap error

### ISSUE THREAD

**Issue by ruiyang2015** [snowflakedb/snowflake-connector-python#1378]
Please answer these questions before submitting your issue. Thanks!

1. What version of Python are you using?
Python 3.9

2. What operating system and processor architecture are you using?

Ubuntu

3. What are the component versions in the environment (`pip freeze`)?

snowflake-connector-python==2.8.3

4. What did you do?
cur.execute("select TO_TIMESTAMP('9999-01-01T00:00:59Z') as d1")
df = cur.fetch_pandas_all()
will throw following exception:
```
File \"/usr/local/lib/python3.9/site-packages/pyarrow/pandas_compat.py\", line 1153, in _table_to_blocks\n    result = pa.lib.table_to_blocks(options, block_table, categories,\n  File \"pyarrow/table.pxi\", line 2602, in pyarrow.lib.table_to_blocks\n  File \"pyarrow/error.pxi\", line 100, in pyarrow.lib.check_status\npyarrow.lib.ArrowInvalid: Casting from timestamp[us] to timestamp[ns] would result in out of bounds timestamp: 253402300799000000"}
```

5. What did you expect to see?
fetch_pands_all() should allow all valid timestamp to loaded,  for out of bound timestamp, allow a way to return NaT or max Timestamp value

6. Can you set logging to DEBUG and collect the logs?

   ```
   import logging
   import os

   for logger_name in ('snowflake.connector',):
       logger = logging.getLogger(logger_name)
       logger.setLevel(logging.DEBUG)
       ch = logging.StreamHandler()
       ch.setLevel(logging.DEBUG)
       ch.setFormatter(logging.Formatter('%(asctime)s - %(threadName)s %(filename)s:%(lineno)d - %(funcName)s() - %(levelname)s - %(message)s'))
       logger.addHandler(ch)
   ```

**Comment by sfc-gh-sfan** [comment]
In 2.7.12 release note: 

> Fixed a bug where timestamps fetched as pandas.DataFrame or pyarrow.Table would overflow for the sake of unnecessary precision. In the case where an overflow cannot be prevented a clear error will be raised now.

I think this is the intended behavior, so I would not call it a bug. But your suggestion of providing an option like pandas (https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#invalid-data) for out-of-bound timestamp is interesting. cc @sfc-gh-mkeller

**Comment by sfc-gh-mkeller** [comment]
This is actually working as intended, please take a look at:
```python
 b = cur.execute("select TO_TIMESTAMP('9999-01-01T00:00:59') as d1").fetch_arrow_all()
 b.to_pandas()   # This will cause an exception
 ```
Our code does [exactly this](https://github.com/snowflakedb/snowflake-connector-python/blob/9b0227ecf3b33e46ca6cb873b0a7ac8b29df79a4/src/snowflake/connector/result_batch.py#L671)
 
 After taking a look at the [pyarrow docs](https://arrow.apache.org/docs/python/generated/pyarrow.Table.html#pyarrow.Table.to_pandas) see:
> timestamp_as_object[bool](https://docs.python.org/3/library/stdtypes.html#bltin-boolean-values), default [False](https://docs.python.org/3/library/constants.html#False)
>    Cast non-nanosecond timestamps (np.datetime64) to objects. This is useful if you have timestamps that don’t fit in the normal date range of nanosecond timestamps (1678 CE-2262 CE). If False, all timestamps are converted to datetime64[ns] dtype.

Unfortunately `pyarrow` does not support `error=coerce`.

**Comment by Victsm** [comment]
We are seeing the same issue after upgrading to 2.8.3
Now, if the table contains any out of range invalid timestamp, reading the table using `fetch_pandas_all` would fail with an exception like the following:
```
Casting from timestamp[us, tz=America/New_York] to timestamp[ns] would result in out of bounds timestamp: -30609792000000000"
```
Reading the table using `fetch_arrow_all` would fail with a different exception like the following:
```
Exception: [Snowflake Exception] arrow failed to build table from batches, errorInfo: Schema at index 1 was different: 
```

The change introduced in 2.7.12 seems too restrictive in terms of tolerating invalid records.
It's not always practical to update the invalid timestamps in the table, and these invalid timestamps shouldn't prevent reading the table.

**Comment by sfc-gh-sfan** [comment]
I'm just curious: if this is also considered as invalid result, can you filter these records out in SQL?

**Comment by ruiyang2015** [comment]
usually, we still want to load this row, but if one of the timestamp clumns contains this out of bound value, you are punishing the entire row for not be able load. filter is out of the question for our use case.

**Comment by sfc-gh-sfan** [comment]
Reading the pyarrow doc linked by @sfc-gh-mkeller , I feel we could potentially provide an option so that user can set `timestamp_as_object` to `True` so that the out-of-range values will be as is? Maybe we can just pass it in as kwargs.

**Comment by Victsm** [comment]
@sfc-gh-sfan , we tried monkey-patching the library to pass `timestamp_as_object=True` [here](https://github.com/snowflakedb/snowflake-connector-python/blob/9b0227ecf3b33e46ca6cb873b0a7ac8b29df79a4/src/snowflake/connector/result_batch.py#L671), however reading the table still failed with the exception inside the C++ ArrowTableIterator:
```
Exception: [Snowflake Exception] arrow failed to build table from batches, errorInfo: Schema at index 1 was different: 
```

**Comment by sfc-gh-sfan** [comment]
I see. Without looking into it deeply, my guess is that `reconstructRecordBatches` might also need to be adapted to this change, so that the schema is changed to accept object instead of timestamp.

**Comment by Victsm** [comment]
The issue I saw seems to be related to #1204.
For the above query that can reproduce the issue, the below error message indicates the schema mismatch is due to the arrow record batch containing the invalid timestamp has downscaled accuracy (`timestamp[us]`).
When merging this arrow record batch with the others containing valid timestamp, it triggers a schema mismatch.

```
Exception: [Snowflake Exception] arrow failed to build table from batches, errorInfo: Schema at index 7 was different: 
D1: timestamp[ns] not null
vs
D1: timestamp[us] not null
```

**Comment by sfc-gh-sfan** [comment]
#1204 just changes the symptom, because before that we just get overflowed value. The core issue is that if we set `timestamp_as_object=True`, then the way we reconstruct the batches in `reconstructRecordBatches` needs to change so we don't construct a timestamp column out of it, but just an object column.

**Comment by shacker** [comment]
I'm dealing with SQL data that includes a lot of timestamps in year 9999 . These are not throwing exceptions, but doing some "wraparound" mangling causing them to return in Pandas as year 1816. I see there are ways to tell Pandas to return these as NaT, but by the time my code gets a handle on it the data has already been mangled by the Snowflake to_pandas() call.  Not sure I have any option other than to not use `fetch_pandas()` here.

**Comment by victorfink** [comment]
I'm having the same issue, with `fetch_pandas_batches`. One of my date columns goes from `timestamp[ns]` to `timestamp[us]`, failing the process.
Is there any way to ignore the error and read the row regardless of the timestamp error?

**Comment by yedu12** [comment]
I'm able to run the fetch_arrow_all() without error but, year is getting converted to 1815 in my case.

```
cursor.execute("select TO_TIMESTAMP('9999-01-01T00:00:59') as d1 from table(generator(rowcount => 100000)) UNION ALL select TO_TIMESTAMP('9999-01-01T00:00:59') as d1").fetch_arrow_all().to_pandas()



  | D1
-- | --
1815-03-31 05:57:07.066277376
1815-03-31 05:57:07.066277376
1815-03-31 05:57:07.066277376
1815-03-31 05:57:07.066277376
1815-03-31 05:57:07.066277376
...
1815-03-31 05:57:07.066277376
1815-03-31 05:57:07.066277376
1815-03-31 05:57:07.066277376
1815-03-31 05:57:07.066277376
1815-03-31 05:57:07.066277376

```

**Comment by sskirby** [comment]
Would it be possible to have an option to leave the timestamps at microsecond precision? I'm running into this error as well, but none of my timestamps even specify nanosecond precision and I would still like to be able to get them out of the DB since Snowflake does support large timestamps.

**Comment by sfc-gh-mkeller** [comment]
@sskirby yes, of course. You can export the timestamps for sure if you cast them into strings

**Comment by sskirby** [comment]
> @sskirby yes, of course. You can export the timestamps for sure if you cast them into strings

@sfc-gh-mkeller is there any way to do this without changing the query?

**Comment by emesar** [comment]
With apologies for reviving a dormant thread, I'm also running into a schema mismatch error (as a `pyarrow.lib.ArrowInvalid` exception) when attempting to call `to_arrow_batches` on a Snowpark DataFrame. Has any progress been made on a solution to this issue?
